charms/kfp-persistence/src/charm.py

```diff
@@ -9,191 +9,71 @@
-                                    "verbs": ["get", "list", "watch"],
-                                },
-                                {
-                                    "apiGroups": [""],
-                                    "resources": ["namespaces"],
-                                    "verbs": ["get"],
-                                },
-                            ],
-                        }
-                    ]
+    """Charm for the data persistence application of Kubeflow Pipelines."""
+
+    def __init__(self, *args, **kwargs):
+        """Initialize charm and setup the container."""
+        super().__init__(*args, **kwargs)
+
+        # Charm logic
+        self.charm_executor = CharmReconciler(self)
+
+        # Components
+        self.leadership_gate = self.charm_executor.add(
+            component=LeadershipGateComponent(charm=self, name="leadership-gate"), depends_on=[]
+        )
+
+        self.kfp_api_relation = self.charm_executor.add(
+            component=SdiRelation(charm=self, name="relation:kfp-api", relation_name="kfp-api"),
+            depends_on=[self.leadership_gate],
+        )
+
+        self.persistenceagent_container = self.charm_executor.add(
+            component=PersistenceAgentContainer(
+                charm=self,
+                name="container:persistenceagent",
+                container_name="persistenceagent",
+                service_name="persistenceagent",
+                files_to_push=[],
+                environment={
+                    "KUBEFLOW_USERID_HEADER": "kubeflow-userid",
+                    "KUBEFLOW_USERID_PREFIX": "",
+                    # Upstream defines this in the configmap persistenceagent-config-*
+                    "MULTIUSER": "true",
                 },
-                "containers": [
-                    {
-                        "name": "ml-pipeline-persistenceagent",
-                        "imageDetails": image_details,
-                        "command": [
-                            "persistence_agent",
-                            "--logtostderr=true",
-                            "--namespace=",
-                            "--ttlSecondsAfterWorkflowFinish=86400",
-                            "--numWorker=2",
-                            f"--mlPipelineAPIServerName={kfpapi['service-name']}",
-                        ],
-                        "envConfig": {
-                            "KUBEFLOW_USERID_HEADER": "kubeflow-userid",
-                            "KUBEFLOW_USERID_PREFIX": "",
-                            # Upstream defines this in the configmap persistenceagent-config-*
-                            "MULTIUSER": "true",
-                        },
-                    }
-                ],
-            },
+                # provide function to pebble with which it can get service configuration from
+                # relation
+                service_config_getter=lambda: PesistenceAgentServiceConfig(
```

Comment:
I think `Component` has `inputs_getter` that you could use for this any avoid implementing your own.  So this could be:
```suggestion
                inputs_getter=lambda: PesistenceAgentServiceConfig(
```
and then in get_layer you can access it through `self._inputs_getter`

I wonder though if it feels better doing it the way you've done it here as its really obvious where the data is going

Reply:
You are correct, it makes things less confusing. I might have multiple getters: one for serice config, another for interfaces, and another for inputs (default).

Reply:
I still think we should keep `_inputs_getter` in base class. It shows how things can be passed in. And if it is sufficient for developer `_inputs_getter` will be used.

Reply:
Will change it to use inputs_getter.